.emacs.d/config.org
#+TITLE: Emacs Configuration
#+OPTIONS: toc:nil author:nil num:nil

* Core
** Sane defaults
  #+BEGIN_SRC emacs-lisp
    (setq scroll-conservatively 100)
    (setq-default buffer-file-coding-system 'utf-8-unix)
    (set-default-coding-systems 'utf-8-unix)
    (setq locale-coding-system 'utf-8-unix)
    (prefer-coding-system 'utf-8-unix)
    (setq tramp-default-method "ssh")
    (setq comint-prompt-read-only t)
    (setq gc-cons-threshold 100000000)
    (setq evil-want-integration t)
    (setq display-time-day-and-date t)
    (setq display-time-format "%e %b %y  %H:%M")
    (setq display-time-default-load-average nil)
    (display-time-mode 1) ;; add time
    (setq evil-want-keybinding nil)
    (setq evil-want-C-u-scroll t)
    (load-file "~/.emacs.d/sensible-defaults.el")
    (sensible-defaults/use-all-settings)
    (sensible-defaults/use-all-keybindings)
    (sensible-defaults/backup-to-temp-directory)
    (setq custom-file "~/.emacs.d/custom.el")
    (load custom-file 'noerror)
    (setq delete-by-moving-to-trash t)
    (defun system-move-file-to-trash (file)
      "Use \"trash\" to move FILE to the system trash.
                When using Homebrew, install it using \"brew install trash\"."
      (call-process (executable-find "trash")
                    nil 0 nil
                    file))
    ;; Keep all backup and auto-save files in one directory
    (setq backup-directory-alist '(("." . "~/.emacs.d/.cache/backups")))
    (setq auto-save-file-name-transforms '((".*" "~/.emacs.d/.cache/auto-save-list/" t)))

    (setq make-backup-files nil) ; stop creating backup~ files
    (setq auto-save-default nil) ; stop creating #autosave# files
    (setq create-lockfiles nil)  ; stop creating .# files

    (global-auto-revert-mode t)
    (global-visual-line-mode 1)

    (setq
     inhibit-startup-message t         ; Don't show the startup message
     inhibit-startup-screen t          ; or screen
     cursor-in-non-selected-windows t  ; Hide the cursor in inactive windows

     echo-keystrokes 0.1               ; Show keystrokes right away, don't show the message in the scratch buffer
     initial-scratch-message nil       ; Empty scratch buffer
     sentence-end-double-space nil     ; Sentences should end in one space, come on!
     confirm-kill-emacs 'y-or-n-p      ; y and n instead of yes and no when quitting
     )

    (fset 'yes-or-no-p 'y-or-n-p)      ; y and n instead of yes and no everywhere else

    (electric-pair-mode 1)

    (setq user-full-name "Tony Ampomah"
          user-mail-address "tony@arksolutions.it")

    (save-place-mode t)
    (setq column-number-mode t)
    (setq-default indent-tabs-mode nil)

    (setq epg-gpg-program "gpg")
    (setq auth-sources
          '((:source "~/.authinfo.gpg")))
  #+END_SRC

  Use spaces instead of tabs
  #+BEGIN_SRC emacs-lisp
    (setq-default indent-tabs-mode nil)
    (setq tab-width 2)
    (setq js-indent-level 2)
    (setq css-indent-offset 2)
    (setq-default c-basic-offset 2)
    (setq c-basic-offset 2)
    (setq-default tab-width 2)
    (setq-default c-basic-indent 2)
  #+END_SRC

  Delete shell buffer when I type exit
  #+BEGIN_SRC emacs-lisp
    (defun my-shell-mode-hook ()
      "Custom `shell-mode' behaviours."
      ;; Kill the buffer when the shell process exits.
      (let* ((proc (get-buffer-process (current-buffer)))
             (sentinel (process-sentinel proc)))
        (set-process-sentinel
         proc
         `(lambda (process signal)
            ;; Call the original process sentinel first.
            (funcall #',sentinel process signal)
            ;; Kill the buffer on an exit signal.
            (and (memq (process-status process) '(exit signal))
                 (buffer-live-p (process-buffer process))
                 (kill-buffer (process-buffer process)))))))

    (add-hook 'shell-mode-hook 'my-shell-mode-hook)
  #+END_SRC

** Utility functions
  #+BEGIN_SRC emacs-lisp
    (defun kwamedat/visit-emacs-config ()
      (interactive)
      (find-file "~/.emacs.d/config.org"))

    (defun kwamedat/visit-host-file ()
      (interactive)
      (find-file "/etc/hosts"))

    (defun kwamedat/find-in-notes ()
      (interactive)
      (counsel-find-file "~/Dropbox/org"))

    (defun kwamedat/find-in-documents ()
      (interactive)
      (counsel-find-file "~/Documents"))

    (defun kwamedat/find-projects-in-documents ()
      (interactive)
      (counsel-find-file "~/Documents/1Projects"))

    (defun kwamedat/find-areas-in-documents ()
      (interactive)
      (counsel-find-file "~/Documents/2Areas"))

    (defun kwamedat/find-resources-in-documents ()
      (interactive)
      (counsel-find-file "~/Documents/3Resources"))

    (defun kwamedat/visit-todo-list-file ()
      (interactive)
      (find-file "~/Dropbox/org/todo.org"))

    (defun kwamedat/visit-inbox ()
      (interactive)
      (find-file "~/Dropbox/org/inbox.org"))
  #+END_SRC

** Key bindings
  #+BEGIN_SRC emacs-lisp :results none
    (setq mac-option-modifier 'meta)
    (setq mac-command-modifier 'super)
    (global-set-key (kbd "M-3") '(lambda () (interactive) (insert "#")))
    (global-set-key (kbd "s-o") (kbd "C-x o"))
    (global-set-key (kbd "<escape>")      'keyboard-escape-quit)
    (global-set-key (kbd "s-w") (kbd "C-x 0"))
    (global-set-key (kbd "s-p") 'projectile-find-file)
    (global-set-key (kbd "s-P") 'projectile-persp-switch-project)
    (global-set-key (kbd "s-q") 'save-buffers-kill-emacs)
    (global-set-key (kbd "s-k") 'kill-this-buffer)

    ;; Keybonds
    (global-set-key [(super a)] 'mark-whole-buffer)
    (global-set-key [(super v)] 'yank)
    (global-set-key [(super c)] 'kill-ring-save)
    (global-set-key [(super s)] 'save-buffer)
    (global-set-key [(super l)] 'goto-line)
    (global-set-key [(super w)]
                    (lambda () (interactive) (delete-window)))
    (global-set-key [(super z)] 'undo)


    (global-set-key (kbd "M-x") 'counsel-M-x)
    (global-set-key (kbd "s-y") 'counsel-yank-pop)
    (global-set-key (kbd "C-x C-f") 'counsel-find-file)
    (global-set-key (kbd "s-r") 'counsel-imenu)
    (global-set-key (kbd "s-b") 'counsel-switch-buffer)
    (global-set-key (kbd "s-/") 'comment-line)
    (global-set-key (kbd "s-F") 'counsel-rg)
    (global-set-key (kbd "s-f") 'swiper-isearch)
  #+END_SRC

** =exec-path-from-shell=
  #+BEGIN_SRC emacs-lisp
   (use-package exec-path-from-shell)
   (setq exec-path-from-shell-check-startup-files nil)
   (when (memq window-system '(mac ns))
     (exec-path-from-shell-initialize))
  #+END_SRC

** =evil-mode=
  #+BEGIN_SRC emacs-lisp
        (defun kwamedat/config-evil-leader ()
          (evil-leader/set-leader "<SPC>")
          (evil-leader/set-key
            ;; workspaces
            "<tab><tab>" 'persp-switch
            "<tab>c" 'persp-switch
            "<tab>." 'persp-switch
            "<tab>s" 'persp-switch
            "<tab>n" 'persp-switch
            "<tab>d" 'persp-kill
            "<tab>k" 'persp-kill
            "<tab>r" 'persp-rename
            "<tab>R" 'persp-state-restore
            "<tab>[" 'persp-prev
            "<tab>]" 'persp-next
            "<tab>S" 'persp-state-save
            "<tab>L" 'persp-state-load
            ;; files
            "fR"  'projectile-recentf
            "fdd" 'kwamedat/find-in-documents
            "fdp" 'kwamedat/find-projects-in-documents
            "fda" 'kwamedat/find-areas-in-documents
            "fdr" 'kwamedat/find-resources-in-documents
            "ff"  'find-file
            "fh"  'kwamedat/visit-host-file
            "fl"  'locate
            "fp"  'kwamedat/visit-emacs-config
            "fr"  'recentf-open-files
            "fs"  'save-buffer
            "fS"  'write-file
            "ft"  'kwamedat/visit-todo-list-file
            ;; buffers
            "bb"  'persp-counsel-switch-buffer
            "bB"  'ibuffer
            "bk"  'kill-buffer
            "bd"  'kill-buffer
            ;; search
            "sb" 'swiper-isearch
            "sp" 'projectile-ag
            ;; git
            "gR"  'vc-revert
            "gg"  'magit-status
            "gb"  'magit-branch-checkout
            "gR"  'vc-revert
            "g/"  'magit-dispatch
            "gD"  'magit-file-delete
            "gB"  'magit-blame-addition
            "gC"  'magit-clone
            "gF"  'magit-fetch
            "gL"  'magit-log
            "gS"  'magit-stage-file
            "gU"  'magit-unstage-file
            ;; toggle
            "ta"  'company-mode
            "tf"  'flycheck-mode
            "tl"  'nlinum-mode
            "tt"  'toggle-truncate-lines
            "tv"  'global-visual-line-mode
            ;; projects
            "pp"  'projectile-persp-switch-project
            "pi"  'projectile-invalidate-cache
            "pt"  'projectile-run-vterm
            "pb"  'projectile-switch-to-buffer
            "pB"  'persp-buffers
            ;; notes
            "nf" 'kwamedat/find-in-notes
            "na" 'org-agenda
            "np" 'org-pomodoro
            "nn" 'kwamedat/find-in-notes
            "nc" 'org-capture
            "nt" 'kwamedat/visit-todo-list-file
            "ni" 'kwamedat/visit-inbox
            "nv" 'org-search-view
            ":"  'eval-expression
            "cc"  'compile
            "<SPC>"  'projectile-find-file
            "/r"  'counsel-imenu
            "/sp"  'counsel-rg
            ;; insert
            "ie" 'emoji-cheat-sheet-plus-insert
            ;; window
            "w"  'evil-window-map
            ;; help
            "h"  'help-map
            ;; open application
            "oD" 'docker
            "od" 'docker
            "oc" 'calender
            "o-" 'dired-jump
            "oo" '+macos/reveal-in-finder
            "op" 'treemacs
            "oxp"  'password-store-copy
            "oxf"  'password-store-copy-field
            "oxi"  'password-store-insert
            "oxg"  'password-store-generate
            "oxd"  'password-store-remove
            "oxr"  'password-store-rename
            "oxe"  'password-store-edit
            "om"  'mu4e
            "ot"  'shell-pop
            "oT"  'vterm-toggle
            "x"  'counsel-M-x))


        (defun kwamedat/apply-evil-other-package-configs ()
          "Apply evil-dependent settings specific to other packages."
          ;; PHP
          (evil-define-key 'normal php-mode-map (kbd ", tt") 'phpunit-current-test)
          (evil-define-key 'normal php-mode-map (kbd ", tp") 'phpunit-current-project)
          (evil-define-key 'normal php-mode-map (kbd ", tc") 'phpunit-current-class)
          (evil-define-key 'normal php-mode-map (kbd ", n") 'phpactor-create-new-class)
          (evil-define-key 'normal php-mode-map (kbd ", m") 'phpactor-move-class)
          (evil-define-key 'normal php-mode-map (kbd ", i") 'phpactor-import-class)
          (evil-define-key 'normal php-mode-map (kbd ", f") 'php-search-documentation)
          (evil-define-key 'normal php-mode-map (kbd ", r") 'counsel-imenu)

          ;; Org Mode
          (evil-define-key 'normal org-mode-map (kbd ", c") 'org-ctrl-c-ctrl-c)
          (evil-define-key 'normal org-mode-map (kbd ", t") 'org-todo)
          (evil-define-key 'normal org-mode-map (kbd ", s") 'org-schedule)
          (evil-define-key 'normal org-mode-map (kbd ", d") 'org-deadline)
          (evil-define-key 'normal org-mode-map (kbd ", a") 'org-agenda)
          (evil-define-key 'normal org-mode-map (kbd ", e") 'org-export-dispatch)
          (evil-define-key 'normal org-mode-map (kbd ", n") 'org-narrow-to-element)
          (evil-define-key 'normal org-mode-map (kbd ", w") 'widen)
          (evil-define-key 'normal org-mode-map (kbd ", r") 'org-refile)

          ;; Ledger Mode
          (evil-define-key 'normal ledger-mode-map (kbd ", a") 'ledger-add-transaction)

          ;; Easy window movement
          (evil-define-key 'normal global-map (kbd "C-j") 'evil-window-down)
          (evil-define-key 'normal global-map (kbd "C-h") 'evil-window-left)
          (evil-define-key 'normal global-map (kbd "C-k") 'evil-window-up)
          (evil-define-key 'normal global-map (kbd "C-l") 'evil-window-right)

          ;; Dired
          (evil-define-key 'normal dired-mode-map (kbd "C-e") 'dired-toggle-read-only))


        (defun kwamedat/config-evil ()
          "Configure evil mode."
          (evil-add-hjkl-bindings occur-mode-map 'emacs
            (kbd "/")       'evil-search-forward
            (kbd "n")       'evil-search-next
            (kbd "N")       'evil-search-previous
            (kbd "C-d")     'evil-scroll-down
            (kbd "C-u")     'evil-scroll-up
            (kbd "C-w C-w") 'other-window)

          (define-key evil-visual-state-map [escape] 'keyboard-quit)
          (define-key minibuffer-local-map [escape] 'minibuffer-keyboard-quit)
          (define-key minibuffer-local-ns-map [escape] 'minibuffer-keyboard-quit)
          (define-key minibuffer-local-completion-map [escape] 'minibuffer-keyboard-quit)
          (define-key minibuffer-local-must-match-map [escape] 'minibuffer-keyboard-quit)
          (define-key minibuffer-local-isearch-map [escape] 'minibuffer-keyboard-quit))

        (defmacro define-evil-or-global-key (key def &optional state)
          "Define a key KEY with DEF in an Evil map, or in the global map.

                   If the Evil map for STATE is defined (or `normal' if STATE is not
                   provided) the key will be defined in that map.  Failing that, it will
                   be defined globally.

                   Note that STATE should be provided as an unquoted symbol.

                   This macro provides a way to override Evil mappings in the appropriate
                   Evil map in a manner that is compatible with environments where Evil
                   is not used."
          (let* ((evil-map-name (if state
                                    (concat "evil-" (symbol-name state) "-state-map")
                                  "evil-normal-state-map"))
                 (map (if (boundp (intern evil-map-name))
                          (intern evil-map-name)
                        global-map)))
            `(define-key ,map ,key ,def)))

        (use-package evil
          :ensure t
          :commands (evil-mode evil-define-key)
          :init
          (setq evil-want-C-u-scroll t)
          (setq evil-vsplit-window-right t)
          (setq evil-split-window-below t)

          (setq evil-want-abbrev-expand-on-insert-exit nil)
          :config
          (add-hook 'evil-mode-hook 'kwamedat/config-evil)
          (evil-mode 1)

          (use-package evil-leader
            :ensure t
            :config
            (global-evil-leader-mode)
            (kwamedat/config-evil-leader))

          (use-package evil-surround
            :ensure t
            :config
            (global-evil-surround-mode))

          (use-package evil-commentary
            :ensure t
            :config
            (evil-commentary-mode))

          (use-package evil-indent-textobject
            :ensure t)
          (kwamedat/apply-evil-other-package-configs))

        (use-package evil-collection
          :after evil
          :ensure t)
        (evil-collection-init)

        (with-eval-after-load 'company
          (define-key company-active-map (kbd "RET") #'company-complete-selection)
          (define-key company-active-map (kbd "C-SPC") #'company-manual-begin))
        (define-key evil-insert-state-map (kbd "C-SPC") 'company-manual-begin)
  #+END_SRC

  #+RESULTS:
  : company-manual-begin

*** Relative numbers =nlium-relative=
  #+BEGIN_SRC emacs-lisp
    (setq nlinum-relative-redisplay-delay 0)
    (setq nlinum-relative-current-symbol "->")
    (setq nlinum-relative-offset 0)
    (nlinum-relative-setup-evil)
    (add-hook 'prog-mode-hook 'nlinum-relative-mode)
    (use-package nlinum-relative)
  #+END_SRC
*** =wgrep=
#+BEGIN_SRC emacs-lisp
(use-package wgrep)

(eval-after-load 'grep
  '(define-key grep-mode-map
    (kbd "C-x C-q") 'wgrep-change-to-wgrep-mode))

(eval-after-load 'wgrep
  '(define-key grep-mode-map
    (kbd "C-c C-c") 'wgrep-finish-edit))

(setq wgrep-auto-save-buffer t)
#+END_SRC
*** =folding=
#+BEGIN_SRC emacs-lisp
  (use-package origami)
  (global-origami-mode t)
#+END_SRC
** =projectile=
  #+BEGIN_SRC emacs-lisp
    (setq projectile-enable-caching nil)
    (setq projectile-project-search-path '("~/Documents/3Resources"
                                           "~/Documents/2Areas"
                                           "~/Documents/1Projects"
                                           "~/Dropbox/org"
                                           "~/Desktop"
                                           "~/Downloads"
                                           "~/Repo/3Resources"
                                           "~/Repo/2Areas"
                                           "~/Repo/1Projects"))
    (setq projectile-completion-system 'ivy)
    (setq projectile-enable-caching t)
    (setq projectile-switch-project-action #'projectile-dired)

    (setq projectile-mode-line
          '(:eval (if (file-remote-p default-directory)
                      " Prj[*remote*]"
                    (format " Prj[%s]" (projectile-project-name)))))

    (use-package projectile
      :ensure t
      :init (projectile-mode))

    (add-to-list 'projectile-globally-ignored-directories "vendor" "node_modules")
  #+END_SRC

** =which-key=
  #+BEGIN_SRC emacs-lisp
    (use-package which-key
      :config (which-key-mode))
  #+END_SRC

** =dump-jump=
  #+BEGIN_SRC emacs-lisp
    (use-package dumb-jump
      :config
      (define-key evil-normal-state-map (kbd "M-.") 'dumb-jump-go)
      (setq dumb-jump-selector 'ivy))
  #+END_SRC

** =undo-tree=
  #+BEGIN_SRC emacs-lisp
    ;; (setq undo-tree-auto-save-history t
    ;;       undo-tree-history-directory-alist
    ;;       `(("." . , "~/.emacs.d/.cache/undo")))
    ;; (unless (file-exists-p  "~/.emacs.d/.cache/undo")
    ;;   (make-directory "~/.emacs.d/.cache/undo" ))

    (use-package undo-tree
      :ensure t
      :config
      (progn
        (global-undo-tree-mode)))
  #+END_SRC

* Lang
** =php=
*** =php-mode=
    #+BEGIN_SRC emacs-lisp
      (defun my-php-mode-setup ()
        "My PHP-mode hook."
        (require 'flycheck-phpstan)
        (flycheck-mode t)
        (subword-mode nil))
      (add-hook 'php-mode-hook 'my-php-mode-setup)

      (use-package php-mode
        :ensure t
        :config
        (setq-default phpstan-executable 'phpstan)
        (setq-default php-mode-coding-style 'psr2)
        (setq-default flycheck-phpcs-standard "PSR2")
        (setq php-template-compatibility nil))

      (use-package php-extras :ensure t)
    #+END_SRC
*** =phpunit=
    #+BEGIN_SRC emacs-lisp
     (use-package phpunit :ensure t)
    #+END_SRC

*** =phpactor=
    #+BEGIN_SRC emacs-lisp
     (use-package phpactor :ensure t)
    #+END_SRC

*** =flycheck-phpactor=
#+BEGIN_SRC emacs-lisp
     (use-package flycheck-phpstan :ensure t)
#+END_SRC
;; (add-to-list 'load-path "~/.emacs.d/php-doc-block")
      ;; (require 'php-doc-block)
      ;; (add-hook 'php-mode-hook
      ;;           (lambda ()
      ;;               (local-set-key (kbd "<C-tab>") 'php-doc-block)))
    #+END_SRC
** =web=
   #+BEGIN_SRC emacs-lisp
     (use-package web-mode)
   #+END_SRC

** =scss=
   #+BEGIN_SRC emacs-lisp
     (use-package scss-mode
       :config
       (setq scss-compile-at-save nil))
   #+END_SRC

** =less-css=
   #+BEGIN_SRC emacs-lisp
     (use-package less-css-mode)
   #+END_SRC

** =css=
   #+BEGIN_SRC emacs-lisp
     (use-package css-mode
       :config
       (setq css-indent-offset 2))
   #+END_SRC

** =yaml=
   #+BEGIN_SRC emacs-lisp
     (use-package yaml-mode)
   #+END_SRC

** =restclient=
   #+BEGIN_SRC emacs-lisp
     (use-package restclient)
   #+END_SRC

** =javascript=
   #+BEGIN_SRC emacs-lisp
     (use-package coffee-mode)
     (setq js-indent-level 2)
     (add-hook 'coffee-mode-hook
               (lambda ()
                 (yas-minor-mode 1)
                 (setq coffee-tab-width 2)))
   #+END_SRC

** =json=
   #+BEGIN_SRC emacs-lisp
     (use-package json-mode)
   #+END_SRC

** =markdown=
   #+BEGIN_SRC emacs-lisp
     (use-package markdown-mode
       :commands (markdown-mode gfm-mode)
       :mode (("README\\.md\\'" . gfm-mode)
              ("\\.md\\'" . markdown-mode)
              ("\\.markdown\\'" . markdown-mode))
       :init (setq markdown-command "pandoc --no-highlight"))
   #+END_SRC

** =ledger=
  #+BEGIN_SRC emacs-lisp
    (setq ledger-post-amount-alignment-column 70)
    (use-package ledger-mode)
    (use-package flycheck-ledger)
    (use-package evil-ledger)
  #+END_SRC
** =csv=
#+BEGIN_SRC emacs-lisp
  (use-package csv-mode)
#+END_SRC
** =org=
  Toggle the local fold at the point (as opposed to cycling through all levels with `org-cycle').
  #+BEGIN_SRC emacs-lisp
    ;;;###autoload
    (defun +org-cycle-only-current-subtree-h (&optional arg)
      (interactive "P")
      (unless (eq this-command 'org-shifttab)
        (save-excursion
          (org-beginning-of-line)
          (let (invisible-p)
            (when (and (org-at-heading-p)
                       (or org-cycle-open-archived-trees
                           (not (member org-archive-tag (org-get-tags))))
                       (or (not arg)
                           (setq invisible-p (outline-invisible-p (line-end-position)))))
              (unless invisible-p
                (setq org-cycle-subtree-status 'subtree))
              (org-cycle-internal-local)
              t)))))

  #+END_SRC

  Stop flycheck from linting TODO keyword in =org-mode=
  #+BEGIN_SRC emacs-lisp
    (defun disable-fylcheck-in-org-src-block ()
      (setq-local flycheck-disabled-checkers '(emacs-lisp-checkdoc)))

    (add-hook 'org-src-mode-hook 'disable-fylcheck-in-org-src-block)


  #+END_SRC

  Org mode configuration
  #+BEGIN_SRC emacs-lisp
    (setq org-icalendar-include-todo t
          org-icalendar-use-deadline '(event-if-todo event-if-not-todo todo-due)
          org-icalendar-use-scheduled '(event-if-todo event-if-not-todo todo-start)
          org-icalendar-with-timestamps t)
    (setq org-startup-truncated nil)
    (setq org-agenda-inhibit-startup nil)
    (add-hook 'org-mode-hook 'org-indent-mode)

    (setq org-refile-targets '((org-agenda-files :maxlevel . 2)))

    (setq org-ellipsis "⤵")

    (setq org-src-fontify-natively t)

    (setq org-directory "~/Dropbox/org")
    (setq org-inbox-file "~/Dropbox/org/inbox.org")
    (setq org-agenda-files (quote (
                                   "~/Dropbox/org/todo.org"
                                   "~/Dropbox/org/routine.org"
                                   )))

    (add-hook 'org-capture-mode-hook 'evil-insert-state)
    (setq org-capture-templates
          '(("t" "Task" entry (file "~/Dropbox/org/inbox.org")
             "* TODO %?\n")
            ("p" "Project" entry (file+headline "~/Dropbox/org/todo.org" "1Projects")
             (file "~/Dropbox/org/templates/new-project-template.org"))
            ("g" "Someday" entry (file+headline "~/Dropbox/org/someday.org" "Someday")
             "* SOMEDAY %?\n")
            ("l" "Log" entry (file+olp+datetree "~/Dropbox/org/log.org" "Log")
             (file "~/Dropbox/org/templates/logtemplate.org"))))

    (org-babel-do-load-languages
     'org-babel-load-languages
     '((R . t)
       (emacs-lisp . t)
       (latex . t)
       (shell . t)
       (elasticsearch . t)
       (ledger . t)
       (python . t)
       (ruby . t)
       (sqlite . t)))

    (setq org-confirm-babel-evaluate nil)

    (setq initial-major-mode 'org-mode)
  #+END_SRC

  Install =org-mode=
  #+BEGIN_SRC emacs-lisp
    (use-package org
      :custom-face
      (variable-pitch ((t (:family "Avenir"))))
      (org-document-title ((t (:weight bold ))))
      (org-done ((t (:strike-through t :weight bold))))
      (org-headline-done ((t (:strike-through t))))
      (org-level-1 ((t (:weight bold :height 1.2))))
      (org-level-2 ((t (:weight normal :height 1.2))))
      (org-level-3 ((t (:weight normal :height 1.2))))
      (org-image-actual-width '(600))
      :config
      (add-hook 'org-tab-first-hook #'+org-cycle-only-current-subtree-h 'append)
      (setq org-todo-keywords
      '((sequence "TODO(t)"
                  "STARTED(s)"
                  "WAITING(w)"
                  "SOMEDAY(.)"
                  "MAYBE(m)"
                  "|"
                  "DONE(d!)"
                  "CANCELLED(c)")))
      :ensure t)
  #+END_SRC

  Instal
*** =org-bullets=
#+BEGIN_SRC emacs-lisp
(use-package org-bullets
  :init
  (add-hook 'org-mode-hook 'org-bullets-mode))
#+END_SRC

*** =ox-pandoc=
   #+BEGIN_SRC emacs-lisp
    (use-package ox-pandoc)
   #+END_SRC
*** =ox-jira=
   #+BEGIN_SRC emacs-lisp
    (use-package ox-jira)
   #+END_SRC
*** =ox-slack=
   #+BEGIN_SRC emacs-lisp

.doom.d/config.el
(setq doom-localleader-key ",")
(setq display-line-numbers-type 'relative)
(setq doom-theme 'doom-one)
(add-to-list 'default-frame-alist '(inhibit-double-buffering . t))

(setq company-idle-delay 0.1)
(setq company-minimum-prefix-length 3)

(setq user-full-name "Tony Ampomah"
      user-mail-address "tony@arksolutions.it"

      doom-font (font-spec :family "Fira Code" :size 14)
      doom-big-font (font-spec :family "Fira Code" :size 14)
      doom-variable-pitch-font (font-spec :family "Fira Code" :size 14)

      which-key-idle-delay 0.40
      lsp-ui-sideline-enable nil
      lsp-ui-doc-enable nil
      dart-format-on-save t
      web-mode-markup-indent-offset 2
      web-mode-code-indent-offset 2
      web-mode-css-indent-offset 2
      js-indent-level 2
      typescript-indent-level 2
      json-reformat:indent-width 2
      prettier-js-args '("--single-quote")
      dired-dwim-target t
      css-indent-offset 2)

(auth-source-pass-enable)
(setq auth-sources '((:source "~/.authinfo.gpg")))
(setq lsp-enable-file-watchers nil)

(when IS-LINUX
  (font-put doom-font :weight 'semi-light))
(when IS-MAC
  (setq ns-use-thin-smoothing t)
  (add-hook 'window-setup-hook #'toggle-frame-maximized))

;; keybindings
(map! :m "M-j" #'multi-next-line
      :m "M-k" #'multi-previous-line
      :m "<f5>" #'open-agenda
      :m "§" #'my-org-super-agenda-today

      ;; Easier window movement
      :n "C-h" #'evil-window-left
      :n "C-j" #'evil-window-down
      :n "C-k" #'evil-window-up
      :n "C-l" #'evil-window-right
      "s-r" #'counsel-imenu
      "s-p" #'+ivy/projectile-find-file
      "s-S-p" #'counsel-projectile-switch-project
      "s-S-f" #'+default/search-project

      (:map evil-treemacs-state-map
        "C-h" #'evil-window-left
        "C-l" #'evil-window-right)

      (:when IS-LINUX
        "s-x" #'execute-extended-command
        "s-;" #'eval-expression
        ;; use super for window/frame navigation/manipulation
        "s-w" #'delete-window
        "s-W" #'delete-frame
        "s-n" #'+default/new-buffer
        "s-N" #'make-frame
        "s-q" (if (daemonp) #'delete-frame #'evil-quit-all)
        ;; Restore OS undo, save, copy, & paste keys (without cua-mode, because
        ;; it imposes some other functionality and overhead we don't need)
        "s-z" #'undo
        "s-c" (if (featurep 'evil) #'evil-yank #'copy-region-as-kill)
        "s-v" #'yank
        "s-s" #'save-buffer
        ;; Buffer-local font scaling
        "s-+" #'doom/reset-font-size
        "s-=" #'doom/increase-font-size
        "s--" #'doom/decrease-font-size
        ;; Conventional text-editing keys
        "s-a" #'mark-whole-buffer
        :gi [s-return]    #'+default/newline-below
        :gi [s-S-return]  #'+default/newline-above
        :gi [s-backspace] #'doom/backward-kill-to-bol-and-indent)

      :leader
      (:prefix "f"
        "t" #'+tonyampomah/find-in-dotfiles
        "T" #'+tonyampomah/browse-dotfiles)
      (:prefix "o"
        "m" #'mu4e
        "." #'+pass/edit-entry)
      (:prefix "n"
        "m" #'+tonyampomah/find-notes-for-major-mode
        "p" #'+tonyampomah/find-notes-for-project))

;; lang/php keybinding
(map! :localleader
      :map php-mode-map
      "i" #'phpactor-import-class
      "m" #'phpactor-move-class

      :prefix "f"
      "r" #'lsp-find-references
      "d" #'lsp-describe-thing-at-point

      :prefix "g"
      "d" #'phpactor-goto-definition
      "i" #'phpactor-find-references

      :prefix "t"
      "p" #'phpunit-current-project
      "c" #'phpunit-current-class
      "t" #'phpunit-current-test)
(map! :n "C-]" #'lsp-find-definition)
(global-set-key (kbd "M-3") '(lambda () (interactive) (insert "#")))

;; Org Mode Configuration
(after! org
  (add-to-list 'org-modules 'org-habit t))
(setq org-directory "~/Dropbox/org")
(setq org-agenda-files (quote ("~/Dropbox/org/todo.org"
                               "~/Dropbox/org/inbox.org"
                               "~/Dropbox/org/goals.org"
                               "~/Dropbox/org/calendar/gcal.org")))
;; org capture templates
(setq org-capture-templates
      '(("t" "Task" entry (file "~/Dropbox/org/inbox.org")
         "* TODO %?\n")
        ("p" "Project" entry (file+headline "~/Dropbox/org/todo.org" "Projects")
         (file "~/Dropbox/org/templates/newprojecttemplate.org"))
        ("s" "Goals" entry (file+headline "~/Dropbox/org/goals.org" "Goals")
         "* SOMEDAY %?\n")
        ("l" "Log" entry (file+olp+datetree "~/Dropbox/org/log.org" "Log")
         (file "~/Dropbox/org/templates/logtemplate.org"))))

(setq org-agenda-inhibit-startup nil
      org-agenda-start-on-weekday nil
      org-agenda-skip-deadline-if-done t
      org-agenda-skip-scheduled-if-done t)

(setq-default flycheck-phpcs-standard "PSR2")

;; projectile
(setq projectile-project-search-path '("~/Desktop"
                                       "~/Downloads"
                                       "~/Repo/2Areas"
                                       "~/Repo/1Projects"))

;; ledger
(setq ledger-post-amount-alignment-column 70)

;; configure email
(setq +mu4e-backend 'offlineimap)
(add-to-list 'load-path "/usr/local/share/emacs/site-lisp/mu/mu4e")
(set-email-account! "arksolutions.it"
                    '((mu4e-sent-folder       . "/arksolutions.it/Sent")
                      (mu4e-drafts-folder     . "/arksolutions.it/Drafts")
                      (mu4e-trash-folder      . "/arksolutions.it/Trash")
                      (mu4e-refile-folder     . "/arksolutions.it/INBOX")
                      (smtpmail-smtp-user     . "tony@arksolutions.it")
                      (user-mail-address      . "tony@arksolutions.it")
                      (mu4e-compose-signature . "---\nTony Ampomah"))
                    t)
(after! mu4e
  ;; load package to be able to capture emails for GTD
  (require 'org-mu4e)
  ;; do not use rich text emails
  (remove-hook! 'mu4e-compose-mode-hook #'org-mu4e-compose-org-mode)

  ;; configure mu4e options
  (setq mu4e-confirm-quit nil ; quit without asking
        mu4e-attachment-dir "~/Downloads"
        mu4e-get-mail-command "offlineimap"
        mu4e-user-mail-address-list (list "tony@arksolutions.it" "tony.ampomah@netsells.co.uk"))

  (setq mu4e-bookmarks
        '( ("flag:unread AND NOT flag:trashed"                "Unread messages"        ?u)
           ("date:today..now"                                 "Today's messages"       ?t)
           ("date:7d..now"                                    "Last 7 days"            ?w)
           ("maildir:/sent"                                   "sent"                   ?s)
           ("maildir:/arksolutions.it/INBOX AND date:7d..now" "Personal Last 7 days"   ?p)
           ("maildir:/netsells.co.uk/INBOX AND date:7d..now"  "Work Last 7 days"       ?n)
           ("mime:image/*"                                    "Messages with images"   ?i)))

  (setq mu4e-maildir-shortcuts
        '( ("/netsells.co.uk/INBOX" . ?w)
           ("/arksolutions.it/INBOX"       . ?p)
           ("/sent"            . ?s)))

  (setq message-send-mail-function 'smtpmail-send-it
        smtpmail-stream-type 'starttls
        smtpmail-default-smtp-server "smtp.gmail.com"
        smtpmail-smtp-server "smtp.gmail.com"
        smtpmail-smtp-service 587)
  ;; add custom actions for messages
  (add-to-list 'mu4e-view-actions
	             '("View in browser" . mu4e-action-view-in-browser) t))
